backend/server/src/database/models.rs

```diff
@@ -428,15 +436,82 @@ impl Campaign {
     }
 
     /// return all campaigns that are live to all users
-    pub fn get_all_public(conn: &PgConnection) -> Vec<Campaign> {
+    pub fn get_all_public_with_roles(conn: &PgConnection, user_id: i32) -> Vec<CampaignWithRoles> {
         use crate::database::schema::campaigns::dsl::*;
 
         let now = Utc::now().naive_utc();
-        campaigns
-            .filter(starts_at.ge(now).or(draft.eq(false)))
+        let campaigns_vec: Vec<Campaign> = campaigns
+            .filter(starts_at.lt(now).and(published.eq(true)))
             .order(id.asc())
             .load(conn)
-            .unwrap_or_else(|_| vec![])
+            .unwrap_or_else(|_| vec![]);
+
+        let mut current_campaigns_roles_vec: Vec<CampaignWithRoles> = Vec::new();
+        for campaign in campaigns_vec {
+            let campaign_roles = Role::get_all_from_campaign_id(conn, campaign.id);
+            let mut applied_for: Vec<i32> = Vec::new();
+
+            for role in &campaign_roles {
+                if Application::get_all_from_role_id(conn, role.id)
+                    .into_iter()
+                    .filter(|app| app.user_id == user_id)
+                    .collect::<Vec<Application>>()
```

Comment:
This is stupidly pedantic but I just looked it up and you can use .count() on an iterator instead of collecting it and then using .len() - do what you want with this info

Reply:
i remember trying to find something similar when writing it - willfix :D